### PR TITLE
Rename 'io.jaegertracing.Span' to 'io.jaegertracing.JaegerSpan'

### LINKED ISSUE #434: Rename 'io.jaegertracing.Span' to 'io.jaegertracing.JaegerSpan'
The `Span` class clashes with OpenTracing's `Span`, which is confusing when debugging and reading code like:

```java
public interface Sender {
  int append(Span span) throws SenderException;
}
```

It's also not clear what is part of the public API and what is not.

I suggest renaming the class `io.jaegertracing.Span` to `io.jaegertracing.internal.JaegerSpan`. This class would still extend `io.opentracing.Span`. Usage of `JaegerSpan` would be allowed only by internal components and will not be part of the public API.

This means that the `Span` from the sender example above would be OpenTracing's, as well as `Reporter#report(Span)`.

The same applies to `Tracer` and `SpanContext`.

### PR DESCRIPTION
**Pull request by jpkrohling**
## Which problem is this PR solving?
- Resolves #434

## Short description of the changes
* Rename Jaeger's `Span` to `JaegerSpan`
* Rename Jaeger's `SpanContext` to `JaegerSpanContext`
* Rename Jaeger's `Tracer` to `JaegerTracer`

Not included in this PR are the decisions about what should be public and what should be internal API. This will be handled in #396.

### CODE REVIEW

**Comment by yurishkuro** [comment]
can we avoid the name JaegerBaseSpan?  I don't understand why we cannot use packages for namespacing.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Map;
+
+/**
+ * Represents a Span as seen from Jaeger's perspective. Builds on OpenTracing's {@link Span}, adding properties that
+ * are not part of the standard.
+ *
+ * Should be used wisely by instrumented applications: always prefer OpenTracing's approach when available.
+ */
+public interface JaegerSpan extends Span {
+  Tracer getTracer();
```
Comment (pavolloffay) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: I would prefer not expose tracer here.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  long getStart();
+
+  long getDuration();
+
+  List<Reference> getReferences();
```
Comment (pavolloffay) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: This is more suitable for span context

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+  Map<String, Object> getTags();
+
+  String getOperationName();
+
+  String getServiceName();
```
Comment (pavolloffay) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: This seems more suitable for a tracer

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+ * are not part of the standard.
+ *
+ * Should be used wisely by instrumented applications: always prefer OpenTracing's approach when available.
+ */
+public interface JaegerSpan extends Span {
+  Tracer getTracer();
+
+  long getStart();
+
+  long getDuration();
```
Comment (pavolloffay) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: what are the units? How is defined duration for not finished span?

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java (hunk 1 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: I extracted `JaegerSpan` into an interface based on what other code is consuming. This PR is not about fixing those parts, but rather, in reorganizing the existing classes.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java (hunk 2 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: See my [previous comment](https://github.com/jaegertracing/jaeger-client-java/pull/454#discussion_r196121636)

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java (hunk 3 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: See my [previous comment](https://github.com/jaegertracing/jaeger-client-java/pull/454#discussion_r196121636)

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java (hunk 4 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: The interface does need some javadoc.

**Comment by jpkrohling** [comment]
> I don't see much benefit of JaegerSpan interface. Could you please alaborate why interface and not just exposing public methods?

The interface is what we tell the world that they can rely on. 
JaegerBaseSpan is what we use internally. There might be public methods here which are *not* part of the public interface. For instance, the `getTracer()` method could be part of the `JaegerBaseSpan`, but not `JaegerSpan`.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java (hunk 1 above)
Comment (pavolloffay) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: > I extracted JaegerSpan into an interface based on what other code is consuming. This PR is not about fixing those parts, but rather, in reorganizing the existing classes.

There is a difference to what is needed in this repo(mostly tests) and what is exposed in general.

I don't see a value exposing a tracer in a span as the tracer is needed to start a span.

Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: There's zero difference between what this PR exposes and what the current code exposes, from a consumer's perspective. I'm not exposing anything *more* than before: code that had `io.jaegertracing.Span` before already had access to `getTracer()`.

This PR is really just about reorganizing the structure of the existing classes and is already too big to my taste. I do plan on sending more PRs after this one, moving things we don't intend to expose, such as this `getTracer()` method.

Work on defining the public/internal API will be mostly done as part of #396.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  long getDuration();
+
+  /**
+   * @return a list of {@link Reference} for this span
+   */
+  List<Reference> getReferences();
+
+  /**
+   * @return an immutable map of tags that are part of this span
```
Comment (mabn) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: It's not immutable at the moment - the content can change at any time.

File: jaeger-core/src/main/java/io/jaegertracing/reporters/Reporter.java [hunk 6]
```diff
@@ -14,7 +14,7 @@
 
 package io.jaegertracing.reporters;
 
-import io.jaegertracing.Span;
+import io.opentracing.Span;
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/reporters/Reporter.java]: I don't think this is a good direction. The only OpenTracing facade this library provides is the Tracer API. Reporter and everything else are internal components that should only work with internal span type, and the Tracer is the one responsible for downcasting from OT span to our span. In particular, for reporting we don't even need downcasting, because `JaegerSpan.finish()` calls `tracer.report(this)`, so reporter gets the internal span directly.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java (hunk 5 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpan.java]: Perhaps `unmodifiable` would be a better term, as implementations should return a copy that cannot be modified by the caller (like, `return Collections.unmodifiableMap(tags);`)

**Comment by jpkrohling** [comment]
@pavolloffay said:
> `JaegerTracer` seems to be missing.

I was planning on sending another PR for that change, as this one is already too big to my taste, but as I'm blocked by this PR here, I decided to add another commit with this rename as well.

**Comment by jpkrohling** [comment]
I just had a chat with @yurishkuro and we decided to revert the split between interface and implementation for JaegerSpan. It will be dealt with on #396, part of a bigger scope.

I updated this PR to include this change and also adjusted the PR description.

File: jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java [hunk 7]
```diff
@@ -35,41 +35,42 @@ public BaggageSetter(BaggageRestrictionManager restrictionManager, Metrics metri
...
+  public JaegerSpanContext setBaggage(JaegerSpan jaegerSpan, String key, String value) {
+    Restriction restriction = restrictionManager.getRestriction(jaegerSpan.getServiceName(), key);
     boolean truncated = false;
     String prevItem = null;
     if (!restriction.isKeyAllowed()) {
       metrics.baggageUpdateFailure.inc(1);
-      logFields(span, key, value, prevItem, truncated, restriction.isKeyAllowed());
-      return span.context();
+      logFields(jaegerSpan, key, value, prevItem, truncated, restriction.isKeyAllowed());
+      return ((JaegerSpanContext) jaegerSpan.context());
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java]: cast unnecessary

File: jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java [hunk 8]
```diff
@@ -35,41 +35,42 @@ public BaggageSetter(BaggageRestrictionManager restrictionManager, Metrics metri
...
       value = value.substring(0, restriction.getMaxValueLength());
       metrics.baggageTruncate.inc(1);
     }
-    prevItem = span.getBaggageItem(key);
-    logFields(span, key, value, prevItem, truncated, restriction.isKeyAllowed());
+    prevItem = jaegerSpan.getBaggageItem(key);
+    logFields(jaegerSpan, key, value, prevItem, truncated, restriction.isKeyAllowed());
     metrics.baggageUpdateSuccess.inc(1);
-    return span.context().withBaggageItem(key, value);
+    return ((JaegerSpanContext) jaegerSpan.context()).withBaggageItem(key, value);
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java]: cast unnecessary

File: jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java [hunk 9]
```diff
@@ -35,41 +35,42 @@ public BaggageSetter(BaggageRestrictionManager restrictionManager, Metrics metri
...
   }
 
-  private void logFields(Span span, String key, String value, String prevItem, boolean truncated, boolean valid) {
-    if (!span.context().isSampled()) {
+  private void logFields(JaegerSpan jaegerSpan, String key, String value, String prevItem,
+                         boolean truncated, boolean valid) {
+    if (!((JaegerSpanContext) jaegerSpan.context()).isSampled()) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/baggage/BaggageSetter.java]: cast unnecessary

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java [hunk 10]
```diff
@@ -71,13 +71,15 @@ private B3TextMapCodec(Builder builder) {
 
   @Override
   public void inject(SpanContext spanContext, TextMap carrier) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: as this is internal API, the spanContext should already be Jaeger's

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java [hunk 11]
```diff
@@ -14,9 +14,9 @@
 
 package io.jaegertracing.propagation;
 
-import io.jaegertracing.SpanContext;
-import io.jaegertracing.Tracer;
-import io.jaegertracing.baggage.BaggageRestrictionManager;
+import io.jaegertracing.JaegerSpanContext;
+import io.jaegertracing.JaegerTracer;
+import io.opentracing.SpanContext;
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: this is a semantic change, previously methods were accepting Jaeger context

File: jaeger-core/src/main/java/io/jaegertracing/propagation/CompositeCodec.java [hunk 12]
```diff
@@ -14,7 +14,7 @@
 
 package io.jaegertracing.propagation;
 
-import io.jaegertracing.SpanContext;
+import io.opentracing.SpanContext;
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/CompositeCodec.java]: semantic change

File: jaeger-core/src/main/java/io/jaegertracing/propagation/Extractor.java [hunk 13]
```diff
@@ -14,11 +14,11 @@
 
 package io.jaegertracing.propagation;
 
-import io.jaegertracing.SpanContext;
+import io.opentracing.SpanContext;
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/propagation/Extractor.java]: semantic change. Codecs are an internal API, Tracer takes care of converting the context to Jaeger context.

File: jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java [hunk 14]
```diff
@@ -146,15 +148,15 @@ public void testTracerTagslist() {
   public void testTracerTagslistFormatError() {
     System.setProperty(Configuration.JAEGER_SERVICE_NAME, "Test");
     System.setProperty(Configuration.JAEGER_TAGS, "testTag1, testTag2 = testValue2");
-    io.jaegertracing.Tracer tracer = (io.jaegertracing.Tracer) Configuration.fromEnv().getTracer();
+    JaegerTracer tracer = (JaegerTracer) Configuration.fromEnv().getTracer();
```
Comment (yurishkuro) [inline jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: TBD - I think configuration should be returning JaegerTracer. But it's not part of this PR.

File: jaeger-core/src/test/java/io/jaegertracing/JaegerSpanTest.java [hunk 15]
```diff
@@ -79,56 +80,56 @@ public void testSpanMetrics() {
...
     final BaggageRestrictionManager mgr = Mockito.mock(DefaultBaggageRestrictionManager.class);
-    tracer = new Tracer.Builder(service)
+    tracer = new JaegerTracer.Builder(service)
             .withReporter(reporter)
             .withSampler(new ConstSampler(true))
             .withClock(clock)
             .withBaggageRestrictionManager(mgr)
             .build();
-    span = (Span) tracer.buildSpan("some-operation").start();
+    jaegerSpan = (JaegerSpan) tracer.buildSpan("some-operation").start();
```
Comment (yurishkuro) [inline jaeger-core/src/test/java/io/jaegertracing/JaegerSpanTest.java]: is this cast necessary?

File: jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java (hunk 11 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/propagation/B3TextMapCodec.java]: Reverted.

File: jaeger-core/src/test/java/io/jaegertracing/JaegerSpanTest.java (hunk 15 above)
Comment (jpkrohling) [reply jaeger-core/src/test/java/io/jaegertracing/JaegerSpanTest.java]: Actually, we could mark JaegerTracer's `buildSpan()` to return `JaegerTracer.SpanBuilder`, whose `start()` method would return `JaegerSpan`. This way, callers using a concrete `JaegerTracer` won't need to cast anything...

Comment (jpkrohling) [reply jaeger-core/src/test/java/io/jaegertracing/JaegerSpanTest.java]: Looks like there are other things to fix if we go this route, as `JaegerTracer.SpanBuilder` is not public, so, some tests cannot access it (`NoopMetricsFactoryTest` is the one that complained first).

I'll open an issue to investigate this.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpanContext.java [hunk 16]
```diff
@@ -16,12 +16,13 @@
 
 import io.jaegertracing.exceptions.EmptyTracerStateStringException;
 import io.jaegertracing.exceptions.MalformedTracerStateStringException;
+import io.opentracing.SpanContext;
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerSpanContext.java]: Q: was the refactoring done using IDE type renaming, or via search/replace?  If the latter, it's hard to see if there are any unintended API changes in this diff since the previous type of `SpanContext` is now changed to OT via import, without showing up in the diff.

File: jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java [hunk 17]
```diff
@@ -137,7 +139,7 @@ public void testReporterConfigurationInvalidLogSpans() {
   public void testTracerTagslist() {
     System.setProperty(Configuration.JAEGER_SERVICE_NAME, "Test");
     System.setProperty(Configuration.JAEGER_TAGS, "testTag1=testValue1, testTag2 = testValue2");
-    io.jaegertracing.Tracer tracer = (io.jaegertracing.Tracer) Configuration.fromEnv().getTracer();
+    JaegerTracer tracer = (JaegerTracer) Configuration.fromEnv().getTracer();
```
Comment (yurishkuro) [inline jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: cast still needed?

File: jaeger-core/src/test/java/io/jaegertracing/baggage/BaggageSetterTest.java [hunk 18]
```diff
@@ -53,22 +54,22 @@ public void setUp() throws Exception {
...
     setter = new BaggageSetter(mgr, metrics);
     tracer =
-        new Tracer.Builder(SERVICE)
+        new JaegerTracer.Builder(SERVICE)
             .withReporter(reporter)
             .withSampler(new ConstSampler(true))
             .withMetrics(metrics)
             .build();
-    span = (Span) tracer.buildSpan("some-operation").start();
+    jaegerSpan = (JaegerSpan) tracer.buildSpan("some-operation").start();
```
Comment (yurishkuro) [inline jaeger-core/src/test/java/io/jaegertracing/baggage/BaggageSetterTest.java]: is the cast still necessary?

Comment (jpkrohling) [reply jaeger-core/src/test/java/io/jaegertracing/baggage/BaggageSetterTest.java]: Yes, `tracer.buildSpan()` returns OT's SpanBuilder, which returns OT's Span for the `start()` method. It's basically the same as https://github.com/jaegertracing/jaeger-client-java/pull/454#discussion_r197764641

File: jaeger-core/src/main/java/io/jaegertracing/Configuration.java [hunk 19]
```diff
@@ -217,7 +218,7 @@ public static Configuration fromEnv() {
     return builder;
   }
 
-  public synchronized io.opentracing.Tracer getTracer() {
+  public synchronized Tracer getTracer() {
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/Configuration.java]: Not on this PR: the previous one was OT's Tracer.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerTracer.java [hunk 20]
```diff
@@ -220,23 +223,24 @@ public void close() {
...
     }
 
     @Override
-    public io.opentracing.Tracer.SpanBuilder addReference(
-        String referenceType, io.opentracing.SpanContext referencedContext) {
+    public Tracer.SpanBuilder addReference(String referenceType, SpanContext reference) {
 
-      if (!(referencedContext instanceof SpanContext)) {
+      if (!(reference instanceof JaegerSpanContext)) {
         return this;
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerTracer.java]: https://github.com/jaegertracing/jaeger-client-java/issues/466

File: jaeger-core/src/main/java/io/jaegertracing/JaegerSpanContext.java (hunk 16 above)
Comment (jpkrohling) [reply jaeger-core/src/main/java/io/jaegertracing/JaegerSpanContext.java]: Initially, via the IDE, but I then changed some (most?) places to use OT's SpanContext. Based on our conversation, I later reverted this last step (also manually). The rule of thumb is that there should be no "semantic" change: where it was previously returning OT's Span Context should still return it and vice-versa.

File: jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java (hunk 17 above)
Comment (jpkrohling) [reply jaeger-core/src/test/java/io/jaegertracing/ConfigurationTest.java]: Yes, as `getTracer()` returns OT's Tracer and the test later calls `JaegerTracer#tags()`.

File: jaeger-core/src/main/java/io/jaegertracing/JaegerTracer.java [hunk 21]
```diff
@@ -220,23 +223,24 @@ public void close() {
     }
 
     @Override
-    public io.opentracing.Tracer.SpanBuilder asChildOf(io.opentracing.SpanContext parent) {
+    public Tracer.SpanBuilder asChildOf(SpanContext parent) {
```
Comment (objectiser) [inline jaeger-core/src/main/java/io/jaegertracing/JaegerTracer.java]: Should all the return types on the build be changed to `JaegerTracer.SpanBuilder`?
